pose_lookup/plugin.cpp

```diff
@@ -5,83 +5,127 @@
 #include "common/data_format.hpp"
 #include "common/plugin.hpp"
 
-/*pyh: reusing data_loading from ground_truth_slam*/
+#include "utils.hpp"
 #include "data_loading.hpp"
 
 using namespace ILLIXR;
 
+std::string
+getenv_or(std::string var, std::string default_) {
+        if (std::getenv(var.c_str())) {
+                return {std::getenv(var.c_str())};
+        } else {
+                return default_;
+        }
+}
+
 class pose_lookup_impl : public pose_prediction {
 public:
-    pose_lookup_impl(const phonebook* const pb)
+	pose_lookup_impl(const phonebook* const pb)
 		: sb{pb->lookup_impl<switchboard>()}
 		, _m_sensor_data{load_data()}
 		, dataset_first_time{_m_sensor_data.cbegin()->first}
 		, _m_start_of_time{std::chrono::high_resolution_clock::now()}
 		, _m_vsync_estimate{sb->subscribe_latest<time_type>("vsync_estimate")}
-    {
-    	auto newoffset = correct_pose(_m_sensor_data.begin()->second).orientation;
-    	set_offset(newoffset);
-    }
+                , enable_alignment{bool(std::stoi(getenv_or("ILLIXR_ALIGNMENT_ENABLE", "0")))}
+		// , path_to_alignment{getenv_or("ILLIXR_ALIGNMENT_FILE", "./metrics/alignMatrix.txt")}
+	{
+		if (enable_alignment)
+			load_align_parameters(path_to_alignment, align_rot, align_trans, align_quat, align_scale);
+		// Read position data of the first frame
+		init_pos_offset = _m_sensor_data.cbegin()->second.position;
+
+		auto newoffset = correct_pose(_m_sensor_data.begin()->second).orientation;
+		set_offset(newoffset);
+	}
 
-    virtual fast_pose_type get_fast_pose() const override {
+	virtual fast_pose_type get_fast_pose() const override {
 		const time_type* estimated_vsync = _m_vsync_estimate->get_latest_ro();
 		if(estimated_vsync == nullptr) {
 			std::cerr << "Vsync estimation not valid yet, returning fast_pose for now()" << std::endl;
 			return get_fast_pose(std::chrono::system_clock::now());
 		} else {
 			return get_fast_pose(*estimated_vsync);
 		}
-    }
+	}
 
-    virtual pose_type get_true_pose() const override {
+	virtual pose_type get_true_pose() const override {
 		throw std::logic_error{"Not Implemented"};
-    }
+	}
 
 
-    virtual bool fast_pose_reliable() const override {
+	virtual bool fast_pose_reliable() const override {
 		return true;
-    }
+	}
 
-    virtual bool true_pose_reliable() const override {
+	virtual bool true_pose_reliable() const override {
 		return false;
-    }
+	}
 
 	virtual Eigen::Quaternionf get_offset() override {
-        return offset;
-    }
+		return offset;
+	}
 
 	virtual pose_type correct_pose(const pose_type pose) const override {
 		pose_type swapped_pose;
+		pose_type input_pose;
+
+		// Step 1: Compensate starting point to (0, 0, 0), pos only
```

Comment:
An elegant way to get rid of the data copy here is to make the input pose non-const, but I am not sure if that is an acceptable solution.